java/src/main/native/src/row_conversion.cu

```diff
@@ -594,134 +594,106 @@ __global__ void copy_to_rows(const size_type num_rows, const size_type num_colum
+         relative_row += warp.size()) {
+      auto const absolute_col = relative_col + tile.start_col;
 
-  size_t fetch_index;      //< tile we are currently fetching
-  size_t processing_index; //< tile we are currently processing
-  for (processing_index = fetch_index = 0; processing_index < tiles_remaining; ++processing_index) {
-    // Fetch ahead up to NUM_TILES_PER_KERNEL_LOADED
-    for (; fetch_index < tiles_remaining && fetch_index < (processing_index + stages_count);
-         ++fetch_index) {
-      auto const fetch_tile = tile_infos[blockIdx.x * NUM_TILES_PER_KERNEL_TO_ROWS + fetch_index];
-      auto const num_fetch_cols = fetch_tile.num_cols();
-      auto const num_fetch_rows = fetch_tile.num_rows();
-      auto const num_elements_in_tile = num_fetch_cols * num_fetch_rows;
-      auto const fetch_tile_row_size = fetch_tile.get_shared_row_size(col_offsets, col_sizes);
-      auto const starting_column_offset = col_offsets[fetch_tile.start_col];
-#ifdef ASYNC_MEMCPY_SUPPORTED
-      auto &fetch_barrier = tile_barrier[fetch_index % NUM_TILES_PER_KERNEL_LOADED];
-      // wait for the last use of the memory to be completed
-      if (fetch_index >= NUM_TILES_PER_KERNEL_LOADED) {
-        fetch_barrier.arrive_and_wait();
-      }
-#else
-      // wait for the last use of the memory to be completed
-      if (fetch_index >= NUM_TILES_PER_KERNEL_LOADED) {
-        group.sync();
+      if (input_data[absolute_col] == nullptr || relative_row >= num_tile_rows) {
+        // variable-width data
+        continue;
       }
-#endif // ASYNC_MEMCPY_SUPPORTED
-
-      // to do the copy we need to do n column copies followed by m element copies OR
-      // we have to do m element copies followed by r row copies. When going from column
-      // to row it is much easier to copy by elements first otherwise we would need a running
-      // total of the column sizes for our tile, which isn't readily available. This makes it
-      // more appealing to copy element-wise from input data into shared matching the end layout
-      // and do row-based memcopies out.
-
-      auto const shared_buffer_base = shared[fetch_index % stages_count];
-      for (auto el = static_cast<int>(threadIdx.x); el < num_elements_in_tile; el += blockDim.x) {
-        auto const relative_col = el / num_fetch_rows;
-        auto const relative_row = el % num_fetch_rows;
-        auto const absolute_col = relative_col + fetch_tile.start_col;
-        if (input_data[absolute_col] == nullptr) {
-          // variable-width data
-          continue;
+      auto const absolute_row = relative_row + tile.start_row;
+      auto const col_size = col_sizes[absolute_col];
+      auto const col_offset = col_offsets[absolute_col];
+      auto const relative_col_offset = col_offset - starting_column_offset;
+
+      auto const shared_offset = relative_row * tile_row_size + relative_col_offset;
+      auto const input_src = input_data[absolute_col] + col_size * absolute_row;
```

Comment:
Seems like `col_size`, `col_offset`, `relative_col_offset`, and (above) `absolute_col` could be computed in the outer loop.  Also seems like you could store input_data[absolute_col] from here `auto const input_src = input_data[absolute_col] + col_size * absolute_row;` as a variable in the outer loop.

I'm sure the optimizer will already be doing most of this, but it can't hurt.